File: day78/exist.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def recursion(i, j, k):
            if i < 0 or i == len(board) or j < 0 or j == len(board[i]) or board[i][j] != word[k]: return False
            if k == len(word) - 1: return True
            
            tmp = board[i][j]
            board[i][j] = "#"
            res = recursion(i-1, j, k+1) or recursion(i+1, j, k+1) or recursion(i, j-1, k+1) or recursion(i, j+1, k+1)
            board[i][j] = tmp
            return res

        for i in range(len(board)):
            for j in range(len(board[i])):
                if recursion(i, j, 0):
                    return True
        return False


# class Solution:
#     def exist(self, board: List[List[str]], word: str) -> bool:
#         if not word:
#             return True
#         self.dx = [-1, 1, 0, 0]
#         self.dy = [0, 0, -1, 1]

#         for i in range(len(board)):
#             for j in range(len(board[i])):
#                 if board[i][j] == word[0] and self.recursion(board, word[1:], i, j):
#                     return True
#         return False


#     def recursion(self, board, word, x, y):
#         if not word or len(word) == 0:
#             return True

#         tmp = board[x][y]
#         board[x][y] = "#"
        
#         for t in range(4):
#             xt = x + self.dx[t]
#             yt = y + self.dy[t]
#             if 0 <= xt < len(board) and 0 <= yt < len(board[x]) and board[xt][yt] == word[0] and self.recursion(board, word[1:], xt, yt):
#                 return True
#         board[x][y] = tmp
#         return False

File: day78/exist.java
/*
面试题12. 矩阵中的路径
请设计一个函数，用来判断在一个矩阵中是否存在一条包含某字符串所有字符的路径。路径可以从矩阵中的任意一格开始，每一步可以在矩阵中向左、右、上、下移动一格。如果一条路径经过了矩阵的某一格，那么该路径不能再次进入该格子。例如，在下面的3×4的矩阵中包含一条字符串“bfce”的路径（路径中的字母用加粗标出）。
[["a","b","c","e"],
["s","f","c","s"],
["a","d","e","e"]]
但矩阵中不包含字符串“abfb”的路径，因为字符串的第一个字符b占据了矩阵中的第一行第二个格子之后，路径不能再次进入这个格子。

链接: https://leetcode-cn.com/problems/ju-zhen-zhong-de-lu-jing-lcof/
*/

class Solution {
    
    public char[][] board;
    public String word;
    
    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.word = word;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (dfs(i, j, 0)) return true;
            }
        }
        return false;
    }

    private boolean dfs(int i, int j, int k) {
        if (i < 0 || i == this.board.length || j < 0 || j == this.board[i].length || this.board[i][j] != this.word.charAt(k)) return false;
        if (k == this.word.length() - 1) return true;

        char tmp = this.board[i][j];
        this.board[i][j] = '#';
        boolean res = dfs(i-1, j, k+1) || dfs(i+1, j, k+1) || dfs(i, j-1, k+1) || dfs(i, j+1, k+1);
        this.board[i][j] = tmp;
        return res;
    }

}
